File: module3/2_Backend_Server/path_planner.py
import math
import matplotlib.pyplot as plt

def generate_waypoints(lat0, lon0, heading_deg, length, width, row_spacing=0.6):
    """
    Generates a serpentine waypoint path for the planter.
    Assumes planter starts at Bottom-Right (0,0) and turns LEFT for new rows.
    """
    waypoints = []
    
    # Calculate how many rows we need based on width
    # math.ceil ensures we cover the full width even if it's not perfectly divisible by 0.6
    num_rows = math.ceil(width / row_spacing) + 1
    
    # Convert compass heading to radians
    heading_rad = math.radians(heading_deg)
    
    # Earth's radius in metres (used for flat-earth coordinate translation)
    R = 6378137.0 

    print(f"--- Generating Mission ---")
    print(f"Target: {num_rows} rows, {length}m long.")

    for i in range(num_rows):
        # 1. Local Cartesian Grid (x, y)
        # Planter starts bottom-right. Turning left means moving in the negative X direction.
        x_local = -i * row_spacing
        
        # Determine forward or return pass (Alternating Y coordinates)
        if i % 2 == 0:
            y_start = 0.0
            y_end = float(length)
        else:
            y_start = float(length)
            y_end = 0.0
            
        # We only generate two waypoints per row (the "Long Vector" strategy)
        row_points = [(x_local, y_start), (x_local, y_end)]
        
        for lx, ly in row_points:
            # 2. Rotation Matrix (Aligning local grid to True North)
            # Standard clockwise rotation for geographic compass bearings:
            delta_north = (ly * math.cos(heading_rad)) - (lx * math.sin(heading_rad))
            delta_east = (ly * math.sin(heading_rad)) + (lx * math.cos(heading_rad))
            
            # 3. Geographic Translation (Metres to Lat/Lon)
            lat_new = lat0 + (delta_north / R) * (180.0 / math.pi)
            lon_new = lon0 + (delta_east / (R * math.cos(math.radians(lat0)))) * (180.0 / math.pi)
            
            waypoints.append({
                "row": i,
                "lat": lat_new,
                "lon": lon_new,
                "local_x": lx, # Kept for plotting
                "local_y": ly  # Kept for plotting
            })

    return waypoints

# ==========================================
# UNIT TEST: The "Test-Before-Integrate" Bench
# ==========================================
if __name__ == "__main__":
    # Dummy Start Location (E.g., Federal University of Agriculture, Abeokuta)
    START_LAT = 7.2225
    START_LON = 3.4408
    
    # Farmer Inputs
    INITIAL_HEADING = 45.0  # Facing North-East
    FIELD_LENGTH = 20.0     # 20 metres long
    FIELD_WIDTH = 5.0       # 5 metres wide
    ROW_SPACING = 0.6       # 60 cm
    
    # Run the brain
    mission_plan = generate_waypoints(START_LAT, START_LON, INITIAL_HEADING, FIELD_LENGTH, FIELD_WIDTH, ROW_SPACING)
    
    # Print the first few coordinates to terminal
    for wp in mission_plan[:4]:
        print(f"Row {wp['row']} -> Lat: {wp['lat']:.7f}, Lon: {wp['lon']:.7f}")
        
    # --- VISUALISATION ---
    # We plot the local coordinates to verify the pure geometry
    x_vals = [wp['local_x'] for wp in mission_plan]
    y_vals = [wp['local_y'] for wp in mission_plan]
    
    plt.figure(figsize=(6, 8))
    plt.plot(x_vals, y_vals, marker='o', linestyle='-', color='green', label='Planter Path')
    plt.plot(x_vals[0], y_vals[0], marker='^', color='blue', markersize=12, label='Start (Bottom-Right)')
    plt.plot(x_vals[-1], y_vals[-1], marker='s', color='red', markersize=12, label='End Mission')
    
    plt.title(f"Serpentine Logic Verification ({ROW_SPACING}m spacing)")
    plt.xlabel("Cross-Track (Metres)")
    plt.ylabel("Along-Track (Metres)")
    plt.grid(True)
    plt.legend()
    plt.axis('equal') # Ensures 1m in X looks exactly like 1m in Y
    plt.show()

File: module3/1_Frontend_UI/path_planner.js
/* =============================================================================
   PATH_PLANNER.JS — Boustrophedon Route Generator
   Autonomous Smart Maize Planter — Ground Control Station
   ============================================================================
   Generates a serpentine (boustrophedon) waypoint coverage path for the
   autonomous maize planter.

   Ported from:  2_Backend_Server/path_planner.py  (generate_waypoints)
   Depends on:   geo_math.js  (calculateGlobalCoordinate)
   ============================================================================ */

'use strict';

/**
 * Generate a complete serpentine (boustrophedon) mission plan.
 *
 * The algorithm produces a coverage pattern where the planter:
 *   1. Starts at the bottom-right of the field (local origin 0, 0).
 *   2. Drives along the Y-axis for the full field length (forward pass).
 *   3. Steps LEFT in the negative X direction by one row spacing.
 *   4. Drives back along the Y-axis (return pass — reversed Y coords).
 *   5. Repeats until the full field width is covered.
 *
 * This is the "Long Vector" strategy — only two waypoints per row
 * (start and end), which is sufficient for straight-line autonomous
 * navigation between headlands.
 *
 * For each generated local point, the function calls
 * calculateGlobalCoordinate() from geo_math.js to compute the
 * corresponding WGS-84 latitude and longitude.
 *
 * ── Boustrophedon Pattern (top-down view, 3 rows shown) ──
 *
 *       Row 2         Row 1         Row 0
 *     x=-1.2        x=-0.6         x=0
 *       │             │              │
 *       ▼  ◄──────────┘   ┌─────────┘
 *       │                 │
 *       └──────────────►  ▼
 *
 * @param {number} startLat          — Origin latitude (decimal degrees).
 * @param {number} startLon          — Origin longitude (decimal degrees).
 * @param {number} initialHeadingDeg — Compass heading (degrees, 0 = North).
 * @param {number} fieldLength       — Along-track length of the field (metres).
 * @param {number} fieldWidth        — Cross-track width of the field (metres).
 * @param {number} [rowSpacing=0.6]  — Distance between rows (metres).
 *                                     Defaults to 0.6 m per the planter's
 *                                     mechanical seed spacing.
 *
 * @returns {Array<Object>} An array of waypoint objects, each containing:
 *   {
 *     row:     number,  // Zero-based row index
 *     lat:     number,  // WGS-84 latitude (decimal degrees)
 *     lon:     number,  // WGS-84 longitude (decimal degrees)
 *     local_x: number,  // Cross-track offset (metres, negative = left)
 *     local_y: number   // Along-track offset (metres)
 *   }
 */
function generateWaypoints(startLat, startLon, initialHeadingDeg, fieldLength, fieldWidth, rowSpacing = 0.6) {

  const waypoints = [];

  /* ── Calculate the number of rows ──────────────────────────────────
   * Math.ceil ensures we cover the full field width even when it is
   * not perfectly divisible by the row spacing.
   * The +1 accounts for the first row at x = 0 (fencepost counting). */
  const numRows = Math.ceil(fieldWidth / rowSpacing) + 1;

  console.log(`[PathPlanner] Generating mission — ${numRows} rows, ${fieldLength} m long, ${rowSpacing} m spacing.`);

  for (let i = 0; i < numRows; i++) {

    /* ── 1. Local Cartesian Grid (x, y) ──────────────────────────────
     * The planter starts at the bottom-right (0, 0).
     * Turning LEFT means stepping in the NEGATIVE X direction.
     * Each subsequent row shifts left by one row spacing. */
    const xLocal = -i * rowSpacing;

    /* ── 2. Boustrophedon alternation ────────────────────────────────
     * Even-indexed rows (0, 2, 4 …) → forward pass:  Y goes 0 → length
     * Odd-indexed rows  (1, 3, 5 …) → return pass:   Y goes length → 0
     * This creates the serpentine (back-and-forth) pattern. */
    let yStart, yEnd;
    if (i % 2 === 0) {
      yStart = 0.0;
      yEnd   = fieldLength;
    } else {
      yStart = fieldLength;
      yEnd   = 0.0;
    }

    /* ── 3. Generate two waypoints per row ("Long Vector" strategy) ── */
    const rowPoints = [
      { lx: xLocal, ly: yStart },
      { lx: xLocal, ly: yEnd },
    ];

    for (const point of rowPoints) {
      /* ── 4. Convert local (x, y) → global (lat, lon) ──────────────
       * Delegates to geo_math.js which applies the rotation matrix
       * and flat-earth translation identically to the Python version. */
      const global = calculateGlobalCoordinate(
        startLat,
        startLon,
        point.lx,
        point.ly,
        initialHeadingDeg
      );

      /* ── 5. Assemble the waypoint object ───────────────────────────
       * Structure matches the Pydantic Waypoint model from
       * data_models.py:  { row, lat, lon, local_x, local_y }         */
      waypoints.push({
        row:     i,
        lat:     global.lat,
        lon:     global.lon,
        local_x: point.lx,
        local_y: point.ly,
      });
    }
  }

  console.log(`[PathPlanner] Generated ${waypoints.length} waypoints across ${numRows} rows.`);
  return waypoints;
}
